### PR TITLE
fix: `SegmentHeader` block range is now an `Option<_>`

### PR DESCRIPTION
**Pull request by joshieDo**
When a `SnapshotProviderRW` creates a new file, it commits the configuration before it can be used. And `block_range` was being initialized as `block.start..=block.start`.

However, there is the case that if the node is interrupted before the first commit of a new file, then on restart, the node would return the wrong highest block of the segment. It would return `block.start`, when in fact it's just an empty file.

This would have consequences when calling `get_snapshot_target`, where it would assume that we need `block.start + 1 -> tip` data.

Changing `SegmentHeader` block_range to an option should fix this.

**This will cause existing nodes on `feat/static-files` to be incompatible with this change, since the configuration file format changed**
To make this less painful, I'll create a small cli tool that can generate configuration files (`conf`) with the same ranges, and so they can replace the existing ones. 

--

* Additional fix where we make a consistency heal but weren't committing with `fsync` to disk

### CODE REVIEW

File: crates/primitives/src/snapshot/segment.rs [hunk 1]
```diff
@@ -138,9 +133,11 @@ impl SnapshotSegment {
 /// A segment header that contains information common to all segments. Used for storage.
 #[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
 pub struct SegmentHeader {
-    /// Block range of the snapshot segment
-    block_range: SegmentRangeInclusive,
-    /// Transaction range of the snapshot segment
+    /// Expected block start of the snapshot segment
```
Comment (shekhirin) [inline crates/primitives/src/snapshot/segment.rs:136]: it would be nice to have an explanation in the comment about why it's needed

File: crates/primitives/src/snapshot/segment.rs [hunk 2]
```diff
@@ -272,7 +297,7 @@ impl SegmentHeader {
     /// Returns the row offset which depends on whether the segment is block or transaction based.
     pub fn start(&self) -> u64 {
         match self.segment {
-            SnapshotSegment::Headers => self.block_start(),
+            SnapshotSegment::Headers => self.expected_block_start(),
             SnapshotSegment::Transactions | SnapshotSegment::Receipts => self.tx_start(),
```
Comment (shekhirin) [inline crates/primitives/src/snapshot/segment.rs:301]: i see that `self.tx_start()` panics if `tx_range` is `None`. Shouldn't it be the same for the block range, i.e. we just take `self.block_start()` and panic on `None`?

Don't understand this part of code

File: crates/storage/provider/src/providers/snapshot/manager.rs [hunk 3]
```diff
@@ -200,7 +200,12 @@ impl SnapshotProvider {
         };
 
         jar.delete()?;
-        self.update_index(segment, None)?;
+
+        let mut segment_max_block = None;
+        if !fixed_block_range.start() != 0 {
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/snapshot/manager.rs:205]: 🤔
```suggestion
        if fixed_block_range.start() > 0 {
```

File: crates/storage/nippy-jar/src/writer.rs [hunk 4]
```diff
@@ -119,6 +119,7 @@ impl<'a, H: NippyJarHeader> NippyJarWriter<'a, H> {
         // changes if necessary.
         if !is_created {
             writer.check_consistency_and_heal()?;
+            writer.commit()?;
```
Comment (onbjerg) [inline crates/storage/nippy-jar/src/writer.rs:122]: unrelated fix?

Comment (joshieDo) [reply crates/storage/nippy-jar/src/writer.rs:122]: yes, didn't actually faced any problems, but noticed that we should commit after a potential heal

File: crates/storage/provider/src/providers/snapshot/writer.rs [hunk 5]
```diff
@@ -87,61 +82,63 @@ impl<'a> SnapshotProviderRW<'a> {
...
-
-            let block_start = find_fixed_range(last_block + 1).start();
-            *self.writer.user_header_mut() = SegmentHeader::new(
-                SegmentRangeInclusive::new(block_start, block_start),
-                None,
-                segment,
-            )
-        } else {
-            self.writer.user_header_mut().increment_block()
+        if let Some(last_block) = self.writer.user_header().block_end() {
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/snapshot/writer.rs:119]: can we use expected block range here now instead?

Comment (joshieDo) [reply crates/storage/provider/src/providers/snapshot/writer.rs:119]: yes, but not on this line, on the one below

Comment (shekhirin) [reply crates/storage/provider/src/providers/snapshot/writer.rs:119]: ye

File: crates/primitives/src/snapshot/segment.rs [hunk 6]
```diff
@@ -171,63 +172,67 @@ impl SegmentHeader {
     }
 
     /// Returns the block range.
-    pub fn block_range(&self) -> SegmentRangeInclusive {
-        self.block_range
+    pub fn block_range(&self) -> Option<&SegmentRangeInclusive> {
```
Comment (shekhirin) [inline crates/primitives/src/snapshot/segment.rs:175]: the type is copy, so no ref is needed i think

File: crates/storage/provider/src/providers/snapshot/writer.rs [hunk 7]
```diff
@@ -97,7 +92,13 @@ impl<'a> SnapshotProviderRW<'a> {
...
         self.writer.commit()?;
 
-        self.reader.update_index(self.writer.user_header().segment(), self.get_max_block())?;
+        if let Some(metrics) = &self.metrics {
+            metrics.record_segment_operation(
+                self.writer.user_header().segment(),
+                SnapshotProviderOperation::CommitWriter,
+                Some(start.elapsed()),
+            );
+        }
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/snapshot/writer.rs:101]: nit: move below update so if it fails we don't report the metric

**Review by shekhirin — APPROVED** [review]
LGTM, one metrics nit

File: crates/storage/provider/src/providers/snapshot/writer.rs (hunk 7 above)
Comment (joshieDo) [reply crates/storage/provider/src/providers/snapshot/writer.rs:101]: whether we index correctly or not, the commit itself did go through so i think its a valid metric

**Review by mattsse — APPROVED** [review]
